lms/djangoapps/api_manager/users/views.py

```diff
@@ -1003,17 +1021,51 @@ def get(self, request, user_id, course_id):
             user=student,
             course_id__exact=course_key,
         )
-        current_grade = 0
-        proforma_grade = 0
-        progress_summary = {}
-        grade_summary = {}
-        grading_policy = {}
+
         if len(queryset):
-            current_grade = queryset[0].grade
-            proforma_grade = queryset[0].proforma_grade
-            progress_summary = json.loads(queryset[0].progress_summary)
-            grade_summary = json.loads(queryset[0].grade_summary)
-            grading_policy = json.loads(queryset[0].grading_policy)
+            gradebook_entry = queryset[0]
+            if (gradebook_entry.grade and
+                    gradebook_entry.proforma_grade and
+                    gradebook_entry.progress_summary and
+                    gradebook_entry.grade_summary and
+                    gradebook_entry.grading_policy):
+                current_grade = queryset[0].grade
+                proforma_grade = queryset[0].proforma_grade
+                progress_summary = json.loads(queryset[0].progress_summary)
+                grade_summary = json.loads(queryset[0].grade_summary)
+                grading_policy = json.loads(queryset[0].grading_policy)
+            else:
+                gradebook_values = _recalculate_grade(request, student, course_descriptor)
+                gradebook_entry.grade = gradebook_values["current_grade"]
+                gradebook_entry.proforma_grade = gradebook_values["proforma_grade"]
+                gradebook_entry.progress_summary = json.dumps(gradebook_values["progress_summary"], cls=EdxJSONEncoder)
+                gradebook_entry.grade_summary = json.dumps(gradebook_values["grade_summary"], cls=EdxJSONEncoder)
+                gradebook_entry.grading_policy = json.dumps(gradebook_values["grading_policy"], cls=EdxJSONEncoder)
+                gradebook_entry.save()
```

Comment:
@marjev  this branch causes 500 in progress page as `progress_summary` might be uninitialized:

Traceback (most recent call last):
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic-2.46.0.37/newrelic/hooks/framework_django.py", line 499, in wrapper return wrapped(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/compat.py", line 137, in view return self.dispatch(request, *args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic-2.46.0.37/newrelic/hooks/component_djangorestframework.py", line 27, in _nr_wrapper_APIView_dispatch_ return wrapped(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 77, in wrapped_view return view_func(*args, **kwargs)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 400, in dispatch response = self.handle_exception(exc)
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/rest_framework/views.py", line 397, in dispatch response = handler(request, *args, **kwargs)
File "/edx/app/edxapp/edx-platform/lms/djangoapps/api_manager/users/views.py", line 1071, in get 'courseware_summary': progress_summary,
UnboundLocalError: local variable 'progress_summary' referenced before assignment

Issues (still) blocked: 
* https://edx-wiki.atlassian.net/browse/MCKIN-3494
* https://edx-wiki.atlassian.net/browse/MCKIN-3495